stellar-contract-env-common/src/env_val.rs

```diff
@@ -1,159 +1,220 @@
-use crate::{BitSet, Status, Symbol, Tag};
+use crate::{BitSet, Object, Status, Symbol, Tag, TagType, TaggedVal, Val};
 
 use super::{
-    raw_val::{RawVal, RawValType},
+    raw_val::{RawVal, RawValConvertable},
     xdr::ScObjectType,
-    Env, RawObj,
+    Env,
 };
 use core::{cmp::Ordering, fmt::Debug};
 
-// EnvVal is a RawVal coupled to a specific instance of Env. In the
-// guest we will use this with a zero-sized Guest unit struct,
-// but in the host we provide a Host and Weak<Host> for Env.
+// EnvVal is a RawVal or TaggedVal coupled to a specific instance of Env. In the
+// guest we will use this with a zero-sized Guest unit struct, but in the host
+// we provide a Host and Weak<Host> for Env.
 #[derive(Clone)]
-pub struct EnvVal<E: Env> {
+pub struct EnvVal<E: Env, V: Val> {
     pub env: E,
-    pub val: RawVal,
+    pub val: V,
 }
 
-impl<E: Env> EnvVal<E> {
+impl<E: Env, V: Val> EnvVal<E, V> {
     pub fn env(&self) -> &E {
         &self.env
     }
 }
 
-impl<E: Env> From<EnvVal<E>> for RawVal {
-    #[inline(always)]
-    fn from(ev: EnvVal<E>) -> Self {
-        ev.val
+impl<E: Env, T: TagType> EnvVal<E, TaggedVal<T>> {
+    // This just lets callers disambiguate by name when they call as_ref()
+    // in a context that might want &TaggedVal<T> or &RawRef; it saves
+    // them writing <_ as AsRef<RawVal>>::as_ref(foo)
+    pub fn as_raw_ref(&self) -> &RawVal {
```

Comment:
Given that there is no cost difference between a ref to a val, and a val, we should probably just forego the ref?
```suggestion
    pub fn as_val(&self) -> RawVal {
```
Wdyt?